virtualizarr/zarr.py

```diff
@@ -134,6 +138,80 @@ def replace(
             zarr_format=zarr_format if zarr_format is not None else self.zarr_format,
         )
 
+    def _default_fill_value(self) -> Union[bool, int, float, str, list]:
+        """
+        The value and format of the fill_value depend on the data_type of the array.
+        See here for spec:
+        https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#fill-value
+        """
+        # numpy dtypes's hierarchy lets us avoid checking for all the widths
+        # https://numpy.org/doc/stable/reference/arrays.scalars.html
+        if self.dtype is np.dtype("bool"):
+            return False
+        elif self.dtype is np.dtype("int"):
+            return 0
+        elif self.dtype is np.dtype("float"):
+            return "NaN"
```

Comment:
It seems like the *default* fill value for float is 0.0:

```python
import zarr
import json
store = zarr.store.MemoryStore(mode="w")
z = zarr.empty((1, 1), store=store)
z[:]

array([[0.]])
```

(I'm not sure where on the Array / Store / Other object that information lives.)

It'd be nice if zarr-python had this as a constant that we could reuse. Would that make sense, or is there some reason not to I'm missing?

Reply:
I also see a reference to NaN as a default at https://github.com/zarr-developers/VirtualiZarr/pull/193/files#diff-f5a7b84b3378d903e91ebf06f2db06dca5ad55d12e7c3bf8537a9b9bb1c4cfa0R361 (present on main).

Reply:
I think the specification doesn't require a specific number, just that it not be null. See the note at the bottom of the `fill_value` section https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#fill-value . However `zarr-python` does default to 0 eventually and not NaN https://github.com/zarr-developers/zarr-python/blob/37a8441c20dae3b284803bb1b0d2e6c8f040fb3e/src/zarr/array.py#L231C9-L235C31 . I may have had some trouble with the unit tests, but I think it's better to be as similar as possible to `zarr-python`, I'll change the defaults to 0s.

Reply:
I think this could also be deferred to a later PR, especially if the true solution is to make it clearer what the default is upstream.